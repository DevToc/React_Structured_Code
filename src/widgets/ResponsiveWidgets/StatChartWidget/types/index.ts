import { StatChartType } from '../StatChartWidget.types';
import { DonutStatChart } from './DonutStatChart';
import { HalfDonutStatChart } from './HalfDonutStatChart';
import { ProgressBarStatChart } from './ProgressBarStatChart';
import { IconStatChart } from './IconStatChart';

export const CHART_TYPE_MAP = {
  [StatChartType.Donut]: DonutStatChart,
  [StatChartType.HalfDonut]: HalfDonutStatChart,
  [StatChartType.ProgressBar]: ProgressBarStatChart,
  [StatChartType.Icon]: IconStatChart,
};
