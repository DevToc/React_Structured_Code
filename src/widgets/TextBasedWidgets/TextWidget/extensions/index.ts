export * from './title';
export * from './customKeyboardShortcuts';
export * from './EmptyTextStyle';
