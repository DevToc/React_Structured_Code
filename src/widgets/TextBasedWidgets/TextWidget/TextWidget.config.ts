import StarterKit from '@tiptap/starter-kit';
import TextStyle from '@tiptap/extension-text-style';
import FontFamily from '@tiptap/extension-font-family';
import Underline from '@tiptap/extension-underline';
import { Color } from '@tiptap/extension-color';
import TextAlign from '@tiptap/extension-text-align';
import { FontSize, LineHeight, CustomLink, CustomListItem } from '../common/extensions';

import { Title, CustomKeyboardShortcuts, EmptyTextStyle } from './extensions';

import { TEXT_TAG_TO_STYLES_MAP } from '../common/TextBasedWidgets.config';

import { Key } from '../../../constants/keyboard';

export const KEYBOARD_CONFIG = {
  excludeSpecific: [Key.Enter],
};

export const DEFAULT_TEXT = 'Click to edit text';

const DEFAULT_PM_NODE_ATTRS = {
  lineHeight: '1.2',
  textAlign: 'left',
};

// Default text widget data for each widget in the side menu
export const DEFAULT_TEXT_WIDGET_DATA = {
  title: {
    dimension: {
      width: 433,
      height: 48,
    },
    proseMirrorData: {
      type: 'title',
      attrs: DEFAULT_PM_NODE_ATTRS,
    },
    style: TEXT_TAG_TO_STYLES_MAP.title,
  },
  subtitle: {
    dimension: {
      width: 300,
      height: 34,
    },
    proseMirrorData: {
      type: 'paragraph',
      attrs: DEFAULT_PM_NODE_ATTRS,
    },
    style: TEXT_TAG_TO_STYLES_MAP.subtitle,
  },
  'heading 1': {
    dimension: {
      width: 380,
      height: 43.5,
    },
    proseMirrorData: {
      type: 'heading',
      attrs: {
        ...DEFAULT_PM_NODE_ATTRS,
        level: 1,
      },
    },
    style: TEXT_TAG_TO_STYLES_MAP['heading 1'],
  },
  'heading 2': {
    dimension: {
      width: 315,
      height: 36,
    },
    proseMirrorData: {
      type: 'heading',
      attrs: {
        ...DEFAULT_PM_NODE_ATTRS,
        level: 2,
      },
    },
    style: TEXT_TAG_TO_STYLES_MAP['heading 2'],
  },
  'heading 3': {
    dimension: {
      width: 250,
      height: 29.5,
    },
    proseMirrorData: {
      type: 'heading',
      attrs: {
        ...DEFAULT_PM_NODE_ATTRS,
        level: 3,
      },
    },
    style: TEXT_TAG_TO_STYLES_MAP['heading 3'],
  },
  'heading 4': {
    dimension: {
      width: 215,
      height: 24.5,
    },
    proseMirrorData: {
      type: 'heading',
      attrs: {
        ...DEFAULT_PM_NODE_ATTRS,
        level: 4,
      },
    },
    style: TEXT_TAG_TO_STYLES_MAP['heading 4'],
  },
  'heading 5': {
    dimension: {
      width: 195,
      height: 22,
    },
    proseMirrorData: {
      type: 'heading',
      attrs: {
        ...DEFAULT_PM_NODE_ATTRS,
        level: 5,
      },
    },
    style: TEXT_TAG_TO_STYLES_MAP['heading 5'],
  },
  'heading 6': {
    dimension: {
      width: 170,
      height: 19.5,
    },
    proseMirrorData: {
      type: 'heading',
      attrs: {
        ...DEFAULT_PM_NODE_ATTRS,
        level: 6,
      },
    },
    style: TEXT_TAG_TO_STYLES_MAP['heading 6'],
  },
  paragraph: {
    dimension: {
      width: 164,
      height: 19.5,
    },
    proseMirrorData: {
      type: 'paragraph',
      attrs: DEFAULT_PM_NODE_ATTRS,
    },
    style: TEXT_TAG_TO_STYLES_MAP.paragraph,
  },
};

export const extensions = [
  CustomLink,
  StarterKit.configure({
    heading: {
      levels: [1, 2, 3, 4, 5, 6],
    },
    listItem: false,
  }),
  CustomListItem,
  TextStyle,
  FontFamily,
  FontSize,
  Underline,
  Color,
  ...[LineHeight, TextAlign].map((extension) =>
    extension.configure({
      types: ['heading', 'paragraph', 'title'],
    }),
  ),
  Title,
  CustomKeyboardShortcuts,
  EmptyTextStyle,
];
