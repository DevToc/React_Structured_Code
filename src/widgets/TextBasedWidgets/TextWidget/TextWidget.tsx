import '../common/prosemirror.css';
import { ReactElement, useCallback, useRef, useState, useEffect, Suspense, FC } from 'react';
import { Box, Spinner } from '@chakra-ui/react';
import isEqual from 'lodash.isequal';
import { EditorView } from 'prosemirror-view';
import { useEditor, EditorContent, Editor } from '@tiptap/react';

import { useResizeObserver } from 'hooks/useResizeObserver';
import { Key } from 'constants/keyboard';
import { useEditor as useInfographEditor, useWidget, WidgetBase, ReadOnlyWidgetBase, WidgetToolbar } from 'widgets/sdk';
import { usePageManager } from 'modules/Editor/components/PageManager/usePageManager';

import { ensurePreviousMarks } from '../common/TextBasedWidgets.helpers';
import { WidgetState } from '../common/TextBasedWidgets.types';
import { TextEditorProps, TextWidgetData } from './TextWidget.types';
import { TextWidgetControlKeyboardShortcuts } from './TextWidgetControlKeyboardShortcuts';
import { useTextBoundingBox } from './useTextBoundingBox';
import { DEFAULT_TEXT, extensions } from './TextWidget.config';
import { handlePaste } from './TextWidget.helpers';
import { useDynamicImport } from 'hooks/useDynamicImport';
import { WidgetComponentsMap } from 'widgets/components.lazy';

interface TextWidgetToolbarProps {
  widgetState: WidgetState;
  editor: Editor;
  dispatchUpdateWidget: (options?: Partial<TextWidgetData>) => void;
}

const TextWidgetToolbar: FC<TextWidgetToolbarProps> = (props): ReactElement => {
  const Component = useDynamicImport('textWidgetToolbar', WidgetComponentsMap);

  return (
    <Suspense fallback={<Spinner />}>
      <Component {...props} />
    </Suspense>
  );
};

export const TextWidget = (): ReactElement => {
  const [widgetState, setNewWidgetState] = useState(WidgetState.Default);
  const { textTag, proseMirrorData, updateWidget } = useWidget<TextWidgetData>();
  const {
    isWidgetSelected,
    enableEditorKeyboardShortcuts,
    disableEditorKeyboardShortcuts,
    isWidgetFocused,
    setWidgetFocus,
    boundingBox,
  } = useInfographEditor();

  const isWidgetHighlighted = useRef<boolean>(false);
  const isEditable = widgetState === WidgetState.Edit;
  const editor = useEditor({
    extensions,
    editorProps: {
      handleDOMEvents: {
        // Prevent text drag and drop
        drop: (_, e: Event) => {
          e.preventDefault();
          return false;
        },
      },
      handlePaste(this, view, _, slice) {
        return handlePaste(view, slice, textTag);
      },
      handleTextInput(view: EditorView, from: number, to: number, text: string) {
        return ensurePreviousMarks(view, textTag, from, to, text);
      },
    },
    content: proseMirrorData,
    editable: isEditable,
  });

  const dispatchUpdateWidget = useCallback(
    (options?: Partial<TextWidgetData>) => {
      const newProseMirrorData = editor?.getJSON();

      if (!newProseMirrorData) return;
      if (isEqual(newProseMirrorData, proseMirrorData)) {
        // the height can be updated even if the content is the same
        setTimeout(() => boundingBox.updateRect(), 0);
        return;
      }

      const editorHeight = editor?.view.dom.clientHeight ?? 0;
      const widgetHeight = editorHeight;

      updateWidget({ proseMirrorData: newProseMirrorData, heightPx: widgetHeight, ...options });
      // update the bounding box rect since the height can be updated
      setTimeout(() => boundingBox.updateRect(), 0);
    },
    [updateWidget, editor, proseMirrorData, boundingBox],
  );

  // update the widget state (default, active, edit)
  // Update the keyboard, focus, prosemirror editor as needed for each state change
  const setWidgetState = useCallback(
    (newWidgetState: WidgetState) => {
      setNewWidgetState(newWidgetState);

      const isActive = newWidgetState === WidgetState.Active;
      const isDefault = newWidgetState === WidgetState.Default;
      const isEditable = newWidgetState === WidgetState.Edit;

      // When the widget is in edit state the editor is focused
      // Reset the focus to the widget when state changes from edit -> active for widget keyboard shortcuts to work
      if (isActive && !isWidgetFocused()) setWidgetFocus();

      if (isActive || isDefault) {
        // The editor should never be editable when the widget is not in edit state
        if (editor?.isEditable) editor?.setEditable(false);

        // keyboard shortcuts should be enabled when the widget is not in edit state
        enableEditorKeyboardShortcuts();

        // if the text is empty, set it to the default text when exiting edit mode
        if (editor?.isEmpty) {
          editor.chain().applyEmptyTextStyle().insertContent(DEFAULT_TEXT).run();
          dispatchUpdateWidget();
        }
      }

      if (isEditable) {
        // focus the editor if its not focused
        if (editor && !editor.isFocused) {
          editor.setEditable(true);
          editor.commands.focus('all');
        }

        // disable keyboard shortcuts when the widget is in edit state
        // The duplicate shortcut is still allowed
        disableEditorKeyboardShortcuts({ allowedKeyboardKeys: [Key.D] });
      }
    },
    [
      isWidgetFocused,
      setWidgetFocus,
      editor,
      disableEditorKeyboardShortcuts,
      enableEditorKeyboardShortcuts,
      dispatchUpdateWidget,
    ],
  );

  const onClick = useCallback(
    (e: React.MouseEvent<HTMLElement>) => {
      // If click target (target) matches widget container (currentTarget), keep it in active state
      // otherwise the text content was clicked so enter edit state
      if (widgetState === WidgetState.Active && e.target !== e.currentTarget) setWidgetState(WidgetState.Edit);
      if (widgetState === WidgetState.Default) {
        if (!isWidgetHighlighted.current) return;
        setWidgetState(WidgetState.Active);
        isWidgetHighlighted.current = true;
      }
    },
    [widgetState, setWidgetState],
  );

  // If the widget is not selected anymore, reset the widget state to default
  useEffect(() => {
    if (!isWidgetSelected && widgetState !== WidgetState.Default) {
      setWidgetState(WidgetState.Default);
      isWidgetHighlighted.current = false;
    }
    // run this effect only when widgetState or isWidgetxSelected changes
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [isWidgetSelected, widgetState]);

  // if the widget is deleted ensure that the keyboard shortcuts are enabled again
  useEffect(() => () => enableEditorKeyboardShortcuts(), [enableEditorKeyboardShortcuts]);

  // This effect updates tiptap editor content if there is change via external action like undo/redo
  // For large document, it may require more optimization later on
  useEffect(() => {
    const shouldResetEditorContent =
      editor &&
      !isWidgetSelected &&
      !editor.isFocused &&
      proseMirrorData &&
      !isEqual(editor.getJSON(), proseMirrorData);
    if (shouldResetEditorContent) {
      editor.commands.setContent(proseMirrorData);
    }
  }, [editor, isWidgetSelected, proseMirrorData]);

  const { customOnResizeStart, customOnResize, customOnResizeEnd, customOnRotateStart, customOnDragEnd } =
    useTextBoundingBox({ setWidgetState, editor, isWidgetSelected });

  return (
    <WidgetBase
      onClick={onClick}
      onResizeStart={customOnResizeStart}
      onResizeEnd={customOnResizeEnd}
      onResize={customOnResize}
      onRotateStart={customOnRotateStart}
      onDragEnd={customOnDragEnd}
    >
      <TextEditor
        dispatchUpdateWidget={dispatchUpdateWidget}
        isWidgetSelected={isWidgetSelected}
        setWidgetState={setWidgetState}
        widgetState={widgetState}
        editor={editor}
      />
    </WidgetBase>
  );
};

const TextEditor = ({
  widgetState,
  setWidgetState,
  isWidgetSelected,
  editor,
  dispatchUpdateWidget,
}: TextEditorProps): ReactElement => {
  const textWrapperRef = useRef<HTMLDivElement>(null);
  const { boundingBox } = useInfographEditor();

  const isEditable = widgetState === WidgetState.Edit;
  const focusEditor = () => editor?.chain().focus();

  const onBlur = useCallback(() => {
    editor?.chain().blur();
    // if the text is empty, set it to the default text when exiting edit mode in setWidgetState
    if (!editor?.isEmpty) dispatchUpdateWidget();
  }, [editor, dispatchUpdateWidget]);

  const fitBoundingBoxToTextWhileTyping = ({ height }: { height: number | undefined }) => {
    if (isEditable && editor && typeof height === 'number') boundingBox.updateRect();
  };

  useResizeObserver({ ref: textWrapperRef.current, onResize: fitBoundingBoxToTextWhileTyping });

  if (!editor) return <></>;

  const stopPropagation = (e: React.KeyboardEvent | React.MouseEvent) => {
    if (isEditable) e.stopPropagation();
  };

  return (
    <>
      <WidgetToolbar>
        <TextWidgetToolbar widgetState={widgetState} dispatchUpdateWidget={dispatchUpdateWidget} editor={editor} />
      </WidgetToolbar>
      {isWidgetSelected && (
        <TextWidgetControlKeyboardShortcuts
          focusEditor={focusEditor}
          setWidgetState={setWidgetState}
          widgetState={widgetState}
        />
      )}
      <Box
        _hover={{ cursor: isEditable ? 'text' : 'default' }}
        h='fit-content'
        w='100%'
        cursor={isEditable ? '' : 'move !important'}
        position='relative'
        onClick={stopPropagation}
        onMouseDown={stopPropagation}
        onBlur={onBlur}
        ref={textWrapperRef}
      >
        <EditorContent editor={editor} />
      </Box>
    </>
  );
};

export const ReadOnlyTextWidget = (): ReactElement => {
  const { proseMirrorData } = useWidget<TextWidgetData>();
  const [CustomLink, ...otherExtensions] = extensions;

  const { disableTabbability } = usePageManager();

  let customExtensions = [
    ...otherExtensions,
    // To fix the issue with links on /pl page not clickable
    // due to the global style on anchor tag from proseMirror
    CustomLink.configure({
      HTMLAttributes: {
        style: 'pointer-events: auto',
      },
    }),
  ];

  if (disableTabbability) {
    customExtensions = [
      ...otherExtensions,
      // To avoid `tabbability` on links from the thumbnail in the page slide view
      // that interfere with the page tab order
      CustomLink.configure({
        HTMLAttributes: {
          tabindex: -1,
        },
      }),
    ];
  }

  const editor = useEditor(
    {
      extensions: customExtensions,
      content: proseMirrorData,
      editable: false,
    },
    [proseMirrorData],
  );

  return (
    <ReadOnlyWidgetBase>
      <Box h='fit-content' w='100%' position='relative'>
        <EditorContent editor={editor} />
      </Box>
    </ReadOnlyWidgetBase>
  );
};
