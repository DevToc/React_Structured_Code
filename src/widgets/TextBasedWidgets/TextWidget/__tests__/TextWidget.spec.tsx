// For accessing the prosemirror element
/* eslint-disable testing-library/no-node-access */
/* eslint-disable testing-library/no-container */

import { fireEvent, screen, act } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { FocusProvider } from 'modules/Editor/components/Focus';
import { Key } from 'constants/keyboard';
import { renderWidget, mockSetActiveWidget, mockLoadWidgetToInfograph } from 'utils/test-utils.test';
import { KeyboardShortcut } from 'modules/Editor/components/KeyboardShortcut';
import { TextWidget } from '../TextWidget';
import { generateDefaultData } from '../TextWidget.helpers';
import { DefaultTextType } from '../TextWidget.types';
import { DEFAULT_TEXT } from '../TextWidget.config';

describe('widgets/TextWidget', () => {
  it('should render element tags correctly', () => {
    // TODO: Current text will generate console.warn, may need to rewrite the tests or update tiptap and prosemirror version
    // @link https://github.com/ueberdosis/tiptap/issues/2846
    // @link https://github.com/ProseMirror/prosemirror-state/blob/f1c2ff98a397aa05b9ccf5ee23642bb7e44f05a7/src/selection.ts#L221

    const widgetId = '003.widget-1';
    const TEXT_TYPES: DefaultTextType[] = [
      'paragraph',
      'subtitle',
      'title',
      'heading 1',
      'heading 2',
      'heading 3',
      'heading 4',
      'heading 5',
      'heading 6',
    ];

    TEXT_TYPES.forEach((type) => {
      const widget = { widgetId, ...generateDefaultData(type) };
      const { asFragment } = renderWidget(<TextWidget />, { widget });

      expect(asFragment()).toMatchSnapshot();
    });
  });

  it('Text in the text widget can be edited', async () => {
    const widgetId = '003.widget-1';
    const widget = { widgetId, ...generateDefaultData('paragraph', '') };
    const { container } = renderWidget(<TextWidget />, { widget });

    mockSetActiveWidget(widgetId);
    const textWidget = screen.getAllByTestId(/widgetbase/)[0];

    // 1st click => widget should be active
    // expect contenteditable to be false
    fireEvent.click(textWidget);
    const proseMirrorEditorEl = container.querySelector('.ProseMirror');
    expect(proseMirrorEditorEl).toHaveAttribute('contenteditable', 'false');

    // 2nd click => widget should be editable
    fireEvent.click(proseMirrorEditorEl);
    expect(proseMirrorEditorEl).toHaveAttribute('contenteditable', 'true');

    // manual focus here required
    // editor?.commands.focus('all') does not seem to work in the test environment
    proseMirrorEditorEl.focus();
    await userEvent.keyboard('{backspace}');
    await userEvent.keyboard('abc');

    expect(proseMirrorEditorEl).toHaveTextContent('abc');

    // widget should move back to non-editable state
    fireEvent.keyDown(document.body, { which: Key.Escape });
    expect(proseMirrorEditorEl).toHaveAttribute('contenteditable', 'false');
  });

  it('should set focus back from editor to widget when going from edit -> active', async () => {
    const widgetId = '003.widget-1';
    const widget = { widgetId, ...generateDefaultData('paragraph', '') };
    const { container } = renderWidget(
      <FocusProvider>
        <TextWidget />
      </FocusProvider>,
      { widget },
    );

    const textWidget = screen.getAllByTestId(/widgetbase/)[0];
    mockSetActiveWidget(widgetId);

    const proseMirrorEditorEl = container.querySelector('.ProseMirror');
    fireEvent.click(proseMirrorEditorEl);
    fireEvent.click(proseMirrorEditorEl);
    // manual focus here required
    // editor?.commands.focus('all') does not seem to work in the test environment
    proseMirrorEditorEl.focus();
    expect(proseMirrorEditorEl).toHaveFocus();

    fireEvent.keyDown(document.body, { which: Key.Escape });
    expect(textWidget).toHaveFocus();
  });

  it('should reset to default text if text is deleted when going away from edit', async () => {
    const widgetId = '003.widget-1';
    const widget = { widgetId, ...generateDefaultData('paragraph', 'hello') };
    const { container } = renderWidget(<TextWidget />, { widget });
    mockLoadWidgetToInfograph(widget);
    mockSetActiveWidget(widgetId);

    const proseMirrorEditorEl = container.querySelector('.ProseMirror');
    expect(proseMirrorEditorEl).toHaveTextContent('hello');
    fireEvent.click(proseMirrorEditorEl);
    fireEvent.click(proseMirrorEditorEl);
    // manual focus here required
    // editor?.commands.focus('all') does not seem to work in the test environment
    proseMirrorEditorEl.focus();
    await userEvent.keyboard('{backspace}');
    expect(proseMirrorEditorEl).toHaveTextContent('');

    fireEvent.keyDown(document.body, { which: Key.Escape });
    expect(proseMirrorEditorEl).toHaveTextContent(DEFAULT_TEXT);

    // should not set the text to empty after undoing
    // manual blur required
    act(() => proseMirrorEditorEl.blur());
    mockSetActiveWidget('');
    fireEvent.keyDown(document.body, { which: Key.Z, metaKey: true });
    expect(proseMirrorEditorEl).toHaveTextContent('hello');
  });

  it('Should update text widget while in active / default mode', async () => {
    const widgetId = '003.widget-1';
    const widget = { widgetId, ...generateDefaultData('paragraph', '') };
    const { container } = renderWidget(<TextWidget />, { widget });
    mockSetActiveWidget(widgetId);
    mockLoadWidgetToInfograph(widget);

    // Toggle bold:
    const boldButton = screen.getByLabelText('Bold Icon');
    fireEvent.click(boldButton);
    const proseMirrorEditor = container.querySelector('.ProseMirror');
    const boldTag = proseMirrorEditor.querySelector('strong');
    expect(boldTag).toBeInTheDocument();
    fireEvent.click(boldButton);
    expect(boldTag).not.toBeInTheDocument();
  });

  it('Undo redo should work while widget is in active/default mode', async () => {
    const widgetId = '003.widget-1';
    const widget = { widgetId, ...generateDefaultData('paragraph', '') };
    const { container } = renderWidget(
      <>
        <KeyboardShortcut />
        <TextWidget />
      </>,
      { widget },
    );
    mockLoadWidgetToInfograph(widget);
    mockSetActiveWidget(widgetId);

    const proseMirrorEditor = container.querySelector('.ProseMirror');

    // Style properties should undo/redo
    const boldButton = screen.getByLabelText('Bold Icon');
    fireEvent.click(boldButton);
    const boldTag = proseMirrorEditor.querySelector('strong');
    expect(boldTag).toBeInTheDocument();
    fireEvent.keyDown(document.body, { which: Key.Z, metaKey: true });
    expect(boldTag).not.toBeInTheDocument();

    mockSetActiveWidget(widgetId);
    fireEvent.click(proseMirrorEditor);
    fireEvent.click(proseMirrorEditor);
    // manual focus here required
    // editor?.commands.focus('all') does not seem to work in the test environment
    proseMirrorEditor.focus();

    await userEvent.keyboard('{backspace}');
    await userEvent.keyboard('abc');
    expect(proseMirrorEditor).toHaveTextContent('abc');

    // set the widget to active
    fireEvent.keyDown(document.body, { which: Key.Escape });
    // blur has to be manually triggered
    fireEvent.blur(proseMirrorEditor);

    fireEvent.keyDown(document.body, { which: Key.Z, metaKey: true });
    expect(proseMirrorEditor).toHaveTextContent(DEFAULT_TEXT);
  });
});
