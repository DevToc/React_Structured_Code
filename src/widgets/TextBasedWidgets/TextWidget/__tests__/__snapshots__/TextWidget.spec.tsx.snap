// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`widgets/TextWidget should render element tags correctly 1`] = `
<DocumentFragment>
  <div
    aria-describedby="003.widget-1-description"
    aria-label="paragraph text widget"
    class="widget-container widget-selecto-target  css-1eee8vp"
    data-testid="widgetbase-003.widget-1"
    id="003.widget-1"
    style="display: block; position: absolute; top: 10px; left: 20px; width: 164px; height: 19.5px; transform: rotate(0deg); z-index: 0; cursor: move;"
    tabindex="-1"
  >
    <div
      class="css-18d03fs"
    >
      <div>
        <div
          class="ProseMirror"
          contenteditable="false"
          translate="no"
        >
          <p
            style="line-height: 1.2"
          >
            <span
              style="font-family: Inter; font-size: 16px; color: #2B2B35"
            >
              Click to edit text
            </span>
          </p>
        </div>
      </div>
    </div>
    <span
      class="css-f8n5zr"
    >
      <p
        class="chakra-text css-0"
        id="003.widget-1-description"
      >
        Click to edit text
      </p>
    </span>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`widgets/TextWidget should render element tags correctly 2`] = `
<DocumentFragment>
  <div
    aria-describedby="003.widget-1-description"
    aria-label="paragraph text widget"
    class="widget-container widget-selecto-target  css-1eee8vp"
    data-testid="widgetbase-003.widget-1"
    id="003.widget-1"
    style="display: block; position: absolute; top: 10px; left: 20px; width: 300px; height: 34px; transform: rotate(0deg); z-index: 0; cursor: move;"
    tabindex="-1"
  >
    <div
      class="css-18d03fs"
    >
      <div>
        <div
          class="ProseMirror"
          contenteditable="false"
          translate="no"
        >
          <p
            style="line-height: 1.2"
          >
            <strong>
              <span
                style="font-family: Inter; font-size: 28px; color: #717178"
              >
                Click to edit text
              </span>
            </strong>
          </p>
        </div>
      </div>
    </div>
    <span
      class="css-f8n5zr"
    >
      <p
        class="chakra-text css-0"
        id="003.widget-1-description"
      >
        Click to edit text
      </p>
    </span>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`widgets/TextWidget should render element tags correctly 3`] = `
<DocumentFragment>
  <div
    aria-describedby="003.widget-1-description"
    aria-label="title text widget"
    class="widget-container widget-selecto-target  css-1eee8vp"
    data-testid="widgetbase-003.widget-1"
    id="003.widget-1"
    style="display: block; position: absolute; top: 10px; left: 20px; width: 433px; height: 48px; transform: rotate(0deg); z-index: 0; cursor: move;"
    tabindex="-1"
  >
    <div
      class="css-18d03fs"
    >
      <div>
        <div
          class="ProseMirror"
          contenteditable="false"
          translate="no"
        >
          <p
            data-title="1"
            style="line-height: 1.2"
          >
            <strong>
              <span
                style="font-family: Inter; font-size: 40px; color: #2B2B35"
              >
                Click to edit text
              </span>
            </strong>
          </p>
        </div>
      </div>
    </div>
    <span
      class="css-f8n5zr"
    >
      <p
        class="chakra-text css-0"
        id="003.widget-1-description"
      >
        Click to edit text
      </p>
    </span>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`widgets/TextWidget should render element tags correctly 4`] = `
<DocumentFragment>
  <div
    aria-describedby="003.widget-1-description"
    aria-label="heading 1 text widget"
    class="widget-container widget-selecto-target  css-1eee8vp"
    data-testid="widgetbase-003.widget-1"
    id="003.widget-1"
    style="display: block; position: absolute; top: 10px; left: 20px; width: 380px; height: 43.5px; transform: rotate(0deg); z-index: 0; cursor: move;"
    tabindex="-1"
  >
    <div
      class="css-18d03fs"
    >
      <div>
        <div
          class="ProseMirror"
          contenteditable="false"
          translate="no"
        >
          <h1
            style="line-height: 1.2"
          >
            <strong>
              <span
                style="font-family: Inter; font-size: 36px; color: #2B2B35"
              >
                Click to edit text
              </span>
            </strong>
          </h1>
        </div>
      </div>
    </div>
    <span
      class="css-f8n5zr"
    >
      <p
        class="chakra-text css-0"
        id="003.widget-1-description"
      >
        Click to edit text
      </p>
    </span>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`widgets/TextWidget should render element tags correctly 5`] = `
<DocumentFragment>
  <div
    aria-describedby="003.widget-1-description"
    aria-label="heading 2 text widget"
    class="widget-container widget-selecto-target  css-1eee8vp"
    data-testid="widgetbase-003.widget-1"
    id="003.widget-1"
    style="display: block; position: absolute; top: 10px; left: 20px; width: 315px; height: 36px; transform: rotate(0deg); z-index: 0; cursor: move;"
    tabindex="-1"
  >
    <div
      class="css-18d03fs"
    >
      <div>
        <div
          class="ProseMirror"
          contenteditable="false"
          translate="no"
        >
          <h2
            style="line-height: 1.2"
          >
            <strong>
              <span
                style="font-family: Inter; font-size: 30px; color: #2B2B35"
              >
                Click to edit text
              </span>
            </strong>
          </h2>
        </div>
      </div>
    </div>
    <span
      class="css-f8n5zr"
    >
      <p
        class="chakra-text css-0"
        id="003.widget-1-description"
      >
        Click to edit text
      </p>
    </span>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`widgets/TextWidget should render element tags correctly 6`] = `
<DocumentFragment>
  <div
    aria-describedby="003.widget-1-description"
    aria-label="heading 3 text widget"
    class="widget-container widget-selecto-target  css-1eee8vp"
    data-testid="widgetbase-003.widget-1"
    id="003.widget-1"
    style="display: block; position: absolute; top: 10px; left: 20px; width: 250px; height: 29.5px; transform: rotate(0deg); z-index: 0; cursor: move;"
    tabindex="-1"
  >
    <div
      class="css-18d03fs"
    >
      <div>
        <div
          class="ProseMirror"
          contenteditable="false"
          translate="no"
        >
          <h3
            style="line-height: 1.2"
          >
            <strong>
              <span
                style="font-family: Inter; font-size: 24px; color: #2B2B35"
              >
                Click to edit text
              </span>
            </strong>
          </h3>
        </div>
      </div>
    </div>
    <span
      class="css-f8n5zr"
    >
      <p
        class="chakra-text css-0"
        id="003.widget-1-description"
      >
        Click to edit text
      </p>
    </span>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`widgets/TextWidget should render element tags correctly 7`] = `
<DocumentFragment>
  <div
    aria-describedby="003.widget-1-description"
    aria-label="heading 4 text widget"
    class="widget-container widget-selecto-target  css-1eee8vp"
    data-testid="widgetbase-003.widget-1"
    id="003.widget-1"
    style="display: block; position: absolute; top: 10px; left: 20px; width: 215px; height: 24.5px; transform: rotate(0deg); z-index: 0; cursor: move;"
    tabindex="-1"
  >
    <div
      class="css-18d03fs"
    >
      <div>
        <div
          class="ProseMirror"
          contenteditable="false"
          translate="no"
        >
          <h4
            style="line-height: 1.2"
          >
            <strong>
              <em>
                <span
                  style="font-family: Inter; font-size: 20px; color: #717178"
                >
                  Click to edit text
                </span>
              </em>
            </strong>
          </h4>
        </div>
      </div>
    </div>
    <span
      class="css-f8n5zr"
    >
      <p
        class="chakra-text css-0"
        id="003.widget-1-description"
      >
        Click to edit text
      </p>
    </span>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`widgets/TextWidget should render element tags correctly 8`] = `
<DocumentFragment>
  <div
    aria-describedby="003.widget-1-description"
    aria-label="heading 5 text widget"
    class="widget-container widget-selecto-target  css-1eee8vp"
    data-testid="widgetbase-003.widget-1"
    id="003.widget-1"
    style="display: block; position: absolute; top: 10px; left: 20px; width: 195px; height: 22px; transform: rotate(0deg); z-index: 0; cursor: move;"
    tabindex="-1"
  >
    <div
      class="css-18d03fs"
    >
      <div>
        <div
          class="ProseMirror"
          contenteditable="false"
          translate="no"
        >
          <h5
            style="line-height: 1.2"
          >
            <strong>
              <em>
                <span
                  style="font-family: Inter; font-size: 18px; color: #717178"
                >
                  Click to edit text
                </span>
              </em>
            </strong>
          </h5>
        </div>
      </div>
    </div>
    <span
      class="css-f8n5zr"
    >
      <p
        class="chakra-text css-0"
        id="003.widget-1-description"
      >
        Click to edit text
      </p>
    </span>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`widgets/TextWidget should render element tags correctly 9`] = `
<DocumentFragment>
  <div
    aria-describedby="003.widget-1-description"
    aria-label="heading 6 text widget"
    class="widget-container widget-selecto-target  css-1eee8vp"
    data-testid="widgetbase-003.widget-1"
    id="003.widget-1"
    style="display: block; position: absolute; top: 10px; left: 20px; width: 170px; height: 19.5px; transform: rotate(0deg); z-index: 0; cursor: move;"
    tabindex="-1"
  >
    <div
      class="css-18d03fs"
    >
      <div>
        <div
          class="ProseMirror"
          contenteditable="false"
          translate="no"
        >
          <h6
            style="line-height: 1.2"
          >
            <strong>
              <em>
                <span
                  style="font-family: Inter; font-size: 16px; color: #717178"
                >
                  Click to edit text
                </span>
              </em>
            </strong>
          </h6>
        </div>
      </div>
    </div>
    <span
      class="css-f8n5zr"
    >
      <p
        class="chakra-text css-0"
        id="003.widget-1-description"
      >
        Click to edit text
      </p>
    </span>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
